src/programs/lsm.md

```diff
@@ -0,0 +1,191 @@
+is allowed to happen or not.
+
+Those actions we are talking about are, for example:
+
+* filesystem operations
+  * opening, creating, moving and removing files
+  * mounting and unmounting filesystems
+* task/process operations
+  * allocating and freeing tasks, changing user and group identify for a task
+* socket operations
+  * creating and binding sockets
+  * receiving and sending messages
+
+Each of those actions has a corresponding LSM hook. All LSM hooks are listed in
+the [lsm_hooks.h](https://github.com/torvalds/linux/blob/master/include/linux/lsm_hooks.h)
+header inside the Linux kernel. Each LSM hook usually has some arguments with
+information about each action which should make it easier for modules/programs
+to make a decision, which are listed in the
+[lsm_hook_defs.h](https://github.com/torvalds/linux/blob/master/include/linux/lsm_hook_defs.h)
+header.
+
+Modules or eBPF programs based on LSM are triggered by those hooks and then
+they can decide whether a particular action is allowed to happen or not.
+
+For example, there is an LSM hook named `task_setnice` with the following
+definition:
+
+```c
+LSM_HOOK(int, 0, task_setnice, struct task_struct *p, int nice)
+```
+
+It's getting triggered when a nice value is set for any process in the system.
+If you are not familiar with the "nice value" concept, check out
+[this article](https://en.wikipedia.org/wiki/Nice_(Unix)). This LSM hook comes
+with the following arguments:
+
+* `p` is the instance of `task_struct` which represents the process on which
+  the nice value is set
+* `nice` is the nice value
+
+Developers of module/program attaching to that hook can decide whether
+particular actions of setting a nice value to a process are allowed or not.
+
+eBPF LSM programs have access to all those arguments and also one additional
+argument - `ret` - which is a return value of potential previous eBPF LSM
+programs.
+
+## Writing LSM BPF program
+
+Let's try to create an LSM eBPF program which which is triggered by
+`task_struct` hook. The purpose of this program will be denying to change a
+nice value to lower than 0 (increasing its priority) for a particular process.
+We want to see our eBPF program to make it impossible to execute the following
+command for a particular PID with a negative value:
+
+```bash
+renice [value] -p [pid]
+```
+
+eBPF projects come with two parts: eBPF program(s) and the userspace program.
+To make our example simple, we can try to deny a change of a nice value of
+the userspace process which loads the eBPF program.
+
+The first step is to create a new project:
+
+```bash
+cargo generate --name lsm-nice -d program_type=lsm -d lsm_hook=task_setnice https://github.com/aya-rs/aya-template
+```
+
+That command should create a new Aya project with an empty program attaching to
+the `task_setnice` hook. Let's go to its directory:
+
+```bash
+cd lsm-mount
+```
+
+`task_setnice` has one argument which is an instance of
```

Comment:
One of the arguments passed to the `task_setnice` hook is a pointer to a `task_struct` type [link].